test/SpokePool.Relay.ts

```diff
@@ -325,27 +345,345 @@ describe("SpokePool Relayer Logic", async function () {
+              FillType.FastFill,
+            ]
+          );
+      });
+      it("does not transfer funds if msg.sender is recipient unless its a slow fill", async function () {
+        const _relayData = {
+          ...relayData,
+          // Set recipient == relayer
+          recipient: relayer.address,
+        };
+        const relayExecution = await getRelayExecutionParams(_relayData);
+        await expect(
+          spokePool.connect(relayer).fillRelayUSSInternal(
+            relayExecution,
+            relayer.address,
+            false // isSlowFill
+          )
+        ).to.not.emit(destErc20, "Transfer");
+      });
+      it("sends updatedOutputAmount to updatedRecipient", async function () {
+        const relayExecution = await getRelayExecutionParams(relayData);
+        const _relayExecution = {
+          ...relayExecution,
+          // Overwrite amount to send to be double the original amount
+          updatedOutputAmount: consts.amountToDeposit.mul(2),
+          // Overwrite recipient to depositor which is not the same as the original recipient
+          updatedRecipient: depositor.address,
+        };
+        expect(_relayExecution.updatedRecipient).to.not.equal(relayExecution.updatedRecipient);
+        expect(_relayExecution.updatedOutputAmount).to.not.equal(relayExecution.updatedOutputAmount);
+        await destErc20.connect(relayer).approve(spokePool.address, _relayExecution.updatedOutputAmount);
+        await expect(() =>
+          spokePool.connect(relayer).fillRelayUSSInternal(
+            _relayExecution,
+            relayer.address,
+            false // isSlowFill
+          )
+        ).to.changeTokenBalance(destErc20, depositor, consts.amountToDeposit.mul(2));
+      });
+      it("unwraps native token if sending to EOA otherwise sends wrapped ERC20", async function () {
+        const _relayData = {
+          ...relayData,
+          outputToken: weth.address,
+        };
+        const relayExecution = await getRelayExecutionParams(_relayData);
+        await expect(() =>
+          spokePool.connect(relayer).fillRelayUSSInternal(
+            relayExecution,
+            relayer.address,
+            false // isSlowFill
+          )
+        ).to.changeEtherBalance(recipient, relayExecution.updatedOutputAmount);
+      });
+      it("slow fills send native token out of spoke pool balance", async function () {
+        const _relayData = {
+          ...relayData,
+          outputToken: weth.address,
+        };
+        const relayExecution = await getRelayExecutionParams(_relayData);
+        await weth.connect(relayer).transfer(spokePool.address, relayExecution.updatedOutputAmount);
+        await expect(() =>
+          spokePool.connect(relayer).fillRelayUSSInternal(
+            relayExecution,
+            relayer.address,
+            true // isSlowFill
+          )
+        ).to.changeEtherBalance(recipient, relayExecution.updatedOutputAmount);
+      });
```

Comment:
This one doesn't strictly verify that the SpokePool paid for the slow fill, since it only verifies that the recipient received `updatedOutputAmount`.

nb haven't tested the snippet below.

```suggestion
        const initialSpokeBalance = await weth.connect(relayer).balanceOf(spokePool.address);
        await weth.connect(relayer).transfer(spokePool.address, relayExecution.updatedOutputAmount);
        await expect(() =>
          spokePool.connect(relayer).fillRelayUSSInternal(
            relayExecution,
            relayer.address,
            true // isSlowFill
          )
        ).to.changeEtherBalance(recipient, relayExecution.updatedOutputAmount);
       const spokeBalance = await weth.connect(relayer).balanceOf(spokePool.address);
       expect(spokeBalance.eq(initialSpokeBalance)).to.be.true;
      });
```

Reply:
can use `changeEtherBalances` instead!

Reply:
69db9ef7ecdf6f134c9126e0671580f2c8862b48